File: uva online judge/11504 - Dominos.py
import sys

def read_test_cases():
    for i in range(int(input())):
        n, m = (int(j) for j in input().split())
        adj_list = [[] for j in range(n)]
        for j in range(m):
            u, v = (int(l) for l in input().split())
            adj_list[u - 1].append(v - 1)
        yield n, adj_list

def tarjan(adj_list, v, depth, low, stack, on_stack, d):
    d += 1
    depth[v], low[v], on_stack[v] = d, d, True
    stack.append(v)
    for u in adj_list[v]:
        if depth[u] is None:
            d = tarjan(adj_list, u, depth, low, stack, on_stack, d)
        if on_stack[u] and low[u] < low[v]:
            low[v] = low[u]
    if depth[v] == low[v]:
        while True:
            u = stack.pop()
            on_stack[u] = False
            if v == u:
                break
            low[u] = depth[v]
    return d

def scc_tarjan(n, adj_list):
    d, depth, low, stack, on_stack = -1, [None] * n, [0] * n, [], [False] * n
    for v in range(n):
        if depth[v] is None:
            d = tarjan(adj_list, v, depth, low, stack, on_stack, d)
    cnt, components, in_edges = 0, [False] * n, [False] * n
    for v in range(n):
        if not components[low[v]]:
            cnt += 1
            components[low[v]] = True
        for u in adj_list[v]:
            if not in_edges[low[u]] and low[u] != low[v]:
                cnt -= 1
                in_edges[low[u]] = True
    return cnt

if __name__ == '__main__':
    sys.setrecursionlimit(1000000)
    for n, adj_list in read_test_cases():
        print(scc_tarjan(n, adj_list))

File: uva online judge/11504 - Dominos.c
#include<stdio.h>
#define N 100000

struct AdjacencyList{
    int n, m;
    int vertex[N];
    int edge[N][2];
};

struct Stack{
    int n;
    int array[N];
};

struct AdjacencyList adj_list;

struct Stack stack;

int num[N], low[N], on_stack[N], in_edges[N], components[N];

void read_test_case(struct AdjacencyList *adj_list){
    int i, v, u;
    scanf("%d %d", &adj_list->n, &adj_list->m);
    for(v = 0; v < adj_list->n; v++)
        adj_list->vertex[v] = -1;
    for(i = 0; i < adj_list->m; i++){
        scanf("%d %d", &v, &u);
        adj_list->edge[i][0] = u - 1;
        adj_list->edge[i][1] = adj_list->vertex[v - 1];
        adj_list->vertex[v - 1] = i;
    }
}

void tarjan(struct AdjacencyList *adj_list, int v, int *depth, int num[], int low[], struct Stack *stack, int on_stack[]){
    int i, u;
    num[v] = low[v] = *depth = *depth + 1;
    stack->array[(stack->n)++] = v;
    on_stack[v] = 1;
    for(i = adj_list->vertex[v]; i != -1; i = adj_list->edge[i][1]){
        u = adj_list->edge[i][0];
        if(num[u] == -1)
            tarjan(adj_list, u, depth, num, low, stack, on_stack);
        if(on_stack[u] && low[v] > low[u])
            low[v] = low[u];
    }
    if(num[v] == low[v]){
        while(1){
            u = stack->array[--(stack->n)];
            on_stack[u] = 0;
            if(v == u)
                break;
            low[u] = num[v];
        }
    }
}

int scc_tarjan(struct AdjacencyList *adj_list, int num[], int low[], struct Stack *stack, int on_stack[], int in_edges[N], int components[N]){
    int i, v, u, depth = -1, cnt = 0;
    stack->n = 0;
    for(v = 0; v < adj_list->n; v++){
        num[v] = -1;
        components[v] = in_edges[v] = on_stack[v] = 0;
    }
    for(v = 0; v < adj_list->n; v++)
        if(num[v] == -1)
            tarjan(adj_list, v, &depth, num, low, stack, on_stack);
    for(v = 0; v < adj_list->n; v++){
        if(!components[low[v]]){
            components[low[v]] = 1;
            cnt++;
        }
        for(i = adj_list->vertex[v]; i != -1; i = adj_list->edge[i][1]){
            u = adj_list->edge[i][0];
            if(low[v] != low[u] && !in_edges[low[u]]){
                in_edges[low[u]] = 1;
                cnt--;
            }
        }
    }
    return cnt;
}

int main(){
    int i;
    for(scanf("%d", &i); i > 0; i--){
        read_test_case(&adj_list);
        printf("%d\n", scc_tarjan(&adj_list, num, low, &stack, on_stack, in_edges, components));
    }
}
